### ISSUE TITLE
NullPointerException in ToBeBuiltComputer#doRemoveProject(IProject, IProgressMonitor)

### ISSUE THREAD

**Issue by mor-n4** [eclipse/xtext-eclipse#924]
We recently ran into a `NullPointerException` in `ToBeBuiltComputer#doRemoveProject()` [1]:
```
Caused by: java.lang.NullPointerException
	at org.eclipse.xtext.builder.impl.ToBeBuiltComputer.doRemoveProject(ToBeBuiltComputer.java:203)
	at org.eclipse.xtext.builder.impl.ToBeBuiltComputer.removeProject(ToBeBuiltComputer.java:180)
	...
```

This was caused by method `Storage2UriMapperImpl#getStorages(URI)` returning pairs with `null` as values. These pairs were created by private method `Storage2UriMapperImpl#getStorages(URI, IFile)`, see [2] for the exact location. Probably the client code using these pairs in `ToBeBuiltComputer#doRemoveProject(IProject, IProgressMonitor)` should be made null-safe. For example, by changing [3] to something like
```
onlyOnThisProject = isSameProject
    || storage2Project.getSecond()==null
    || !storage2Project.getSecond().isAccessible();
```

We ran into this issue in our N4JS language, which is using a heavily modified workspace for external libraries, making use of file URIs for special purposes. While I am not sure such a null pointer exception can occur in standard Xtext usage, I think the null check would make sense in any case, because
* it is at least not apparent why this cannot occur in standard, unmodified Xtext usage (at least to me),
* it would be a good fit for the logic already implemented in [3].

Notes:
[1] https://github.com/eclipse/n4js/issues/1190

[2] https://github.com/eclipse/xtext-eclipse/blob/4b25f2ade8ee92cafd77ae9e2434b62ee3baee0c/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/resource/Storage2UriMapperImpl.java#L230-L231

[3] https://github.com/eclipse/xtext-eclipse/blob/4b25f2ade8ee92cafd77ae9e2434b62ee3baee0c/org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/ToBeBuiltComputer.java#L203

**Comment by kthoms** [comment]
Thanks for reporting this issue. At a first glance I think your proposal to change to
```
onlyOnThisProject = isSameProject
    || storage2Project.getSecond()==null
    || !storage2Project.getSecond().isAccessible();
```
makes sense, since "second" can be assumed not to be accessible if it is null. And a NPE should be avoided in any case.

**Comment by kthoms** [comment]
Fixed in 2.17. Thanks for reporting it!
